text/0004-new-pre-coordicide-consensus/0004-new-pre-coordicide-consensus.tex.md

```diff
@@ -0,0 +1,191 @@
+
+
+# Summary
+
+A change to the coordinator based protocol that should enable higher througput of TPS and CTPS. It is achieved by doing the following changes:
+
+1. **White Flag** - Allow milestones to confirm conflicting bundles. Mutate the ledger state by deterministically ordering the bundles and ignoring conflicts.
+2. **New Node Tip Selection** - Nodes doing an almost random tip-selection will guarantee high throughputs and that honest txs will not likely be left behind. By adhering to certain constraints we will ensure that new transactions will add to the cumulative weight of existing transactions. A user will use `getTransactionsToApprove` API call with no parameters to invoke this tip selection.
+3. **New Compass Tip Selection** - Compass will use a heuristic to confirm a suntangle that has passed a certain threshold of cumulative weight (WIP).
+
+# Motivation
+
+The main motivations:
+1. **Defend against censorship attacks** - Conflicts will no longer block bundles from being approved.
+2.  **Make reattachments unneccessary** - As long as the network is not saturated, theoretically all bundles should be approved. And no bundle will be left behind.
+3. **Increase TPS** - Due to easy node tipselection the network throughput should increase.
+4. **Increase CTPS** - Due to the above, increase in TPS and no left-behinds, we expect CTPS to increase as well. 
+
+
+# Detailed design
+
+## White-Flag
+
+Let's define a `conflicting bundle` as a bundle that leads to a negative balance on an address if applied to the current ledger state.
+
+When a milestone is accepted by a node then the following algorithm to mutate the ledger state is performed: 
+
+1. Look for milestone whose index is the `current latest solid milestone index + 1`. Milestones must be applied in order!
+1. When the milestone is found start traversing down the tangle in [DFS](https://en.wikipedia.org/wiki/Depth-first_search) going first to the trunk and only then the branch. Stop at the transactions that has only approved parents.
+2. Start applying bundles to the diffmap only as you go up the recursion stack (when you climb up back from the leafs).
+3. Every time you attempt to apply a bundle to the diffmap:
+   - If it is invalid mark it as `seen` and `ignored` by the milestone. 
+   - If it is valid and not conflicting with the current state then apply it to the ledger state. Mark it as `seen` and `approved`.
+   - If it is valid but conflicting with the current state then place it in an ordered *Conflict Set*.
+4. **Local overdraft handling**: After you are done traversing the subtangle attempt to apply again the conflicting bundles. Iterate the *Conflict Set* in order and attempt to apply the bundles to the ledger state. Mark the applied bundles as `seen` and `approved`. Remove applied bundles from the set. Repeat this step until the *Conflict Set* size doesn't change and mark the remaining bundles of the set as seen and ignored.
```

Comment:
`applied` means that  applying the bundle does not result in negative ledger state?